ibridges/irodsconnector/data_operations.py

```diff
@@ -110,56 +145,252 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+        resc_name : str
+            Name of the resource to which data is uploaded, by default the server will decide
+        options : dict
+            More options for the upload
+        """
+        local_path = Path(local_path)
+        irods_path = IrodsPath(self.session, irods_path)
+        # get all files and their relative path to local_path
+        if not local_path.is_dir():
+            raise ValueError("local_path must be a directory.")
+
+        files = []  # (rel_path, name)
+
+        for w in os.walk(local_path):
+            files.extend([(w[0].removeprefix(str(local_path)), f) for f in w[2]])
+
+        upload_path = IrodsPath(self.session,
+                                str(irods_path.joinpath(local_path.name)))
+        _ = self.create_collection(upload_path)
+
+        for folder, file_name in files:
+            # ensure iRODS collection exists
+            _ = self.create_collection(upload_path.joinpath(folder.lstrip('/')))
+            # call irods put for Path(local_path, f[0], f[1]) to destination collection
+            dest = IrodsPath(self.session, str(upload_path), folder.lstrip('/'))
+            try:
+                self._obj_put(Path(str(local_path), folder.lstrip('/'), file_name),
+                              dest, overwrite, resc_name, options)
+            except irods.exception.OVERWRITE_WITHOUT_FORCE_FLAG:
+                l = Path(str(local_path), folder.lstrip('/'), file_name)
+                warnings.warn(f'Upload: Object already exists\n\tSkipping {l}')
+
+    def _download_collection(self, irods_path: IrodsPath, local_path: Path,
+                             overwrite: bool = False, options: Optional[dict] = None):
+        """Download a collection to the local filesystem
+
+        Parameters
+        ----------
+        irods_path : IrodsPath
+            Absolute irods source path pointing to a collection
+        local_path : Path
+            Absolute path to the destination directory
+        overwrite : bool
+            Overwrite existing local data
+        options : dict
+            More options for the download
+        """
+        irods_path = IrodsPath(self.session, irods_path)
+        if not irods_path.collection_exists():
+            raise ValueError("irods_path must be a collection.")
+
+        # get all data objects
+        coll = self.get_collection(irods_path)
+        all_objs = self._get_data_objects(coll)
+        local_path = local_path.joinpath(irods_path.name.lstrip('/'))
+        if not local_path.is_dir():
+            os.makedirs(local_path)
+
+        for subcoll_path, obj_name, _, _ in all_objs:
+            # ensure local folder exists
+            dest = Path(local_path, subcoll_path.removeprefix(str(irods_path)).lstrip('/'))
+            if not dest.is_dir():
+                os.makedirs(dest)
+            try:
+                self._obj_get(IrodsPath(self.session, subcoll_path, obj_name),
+                              local_path, overwrite, options)
```

Comment:
One last thing: I think `local_path` should be `dest` here. Currently, the subcollection's files end up in the root.

Reply:
Thanks!!!